flowermd/base/simulation.py

```diff
@@ -284,7 +295,16 @@ def volume(self):
     def mass_reduced(self):
         """The total mass of the system in reduced units."""
         with self.state.cpu_local_snapshot as snap:
-            return sum(snap.particles.mass)
+            if self._rigid_constraint:
+                last_body_tag = 0
+                for body_tag in snap.particles.body:
+                    if body_tag == last_body_tag:
+                        last_body_tag += 1
+                    else:
+                        break
```

Comment:
```suggestion
                body = np.array(snap.particles.body)
                last_tag_idx = np.where(body[1:] <body[:-1])[0]
                last_body_tag = last_tag_idx+1 if last_tag_idx else 0
                return sum(snap.particles.mass[last_body_tag:])
                
                
```

Reply:
I wonder if this might not really be faster for large systems, since `np.where` is going to look through all of the elements in snap.particles.body, while the for loop stops once it finds the first case of the body tag not increasing.

Reply:
Yes you're right. I guess I just have this urge to avoid "for" loop :laughing: 

My idea was to check the tags and stop when the ascending order stops. The current way stops the for loop, when the last index and current index are not equal, which implies that the first group of c-particles always belong to first rigid body (or any rigid body with tag < last_body_tag). This works for almost all of our cases now. 
(nitpicking alert!!) It might become an issue in cases where the rigid snapshot is being created in a way that all of sudden the first group of c-particles **actually** belong to the last body. I think this scenario is very rare and it might never happen lol!  Feel free to ignore my comment

Reply:
That's a good point, I'll tweak it so that it is actually checking for ascending tags.